add functions to set a hex's alpha value

musjj opened the pull request:
On Nix, there are some programs that accepts hex values with alpha/transparency (e.g. VS Code, Zed, Firefox). It would be convenient if there's a function for this use-case:

```nix
nix-repl> hex.alpha "#000000" 50
"#0000007F"

nix-repl> hex.alpha "#000000FF" 30
"#0000004C"
```

Let me know what you think.

musjj committed:
add functions to set a hex's alpha value

nutsalhan87 commented:
This is great idea. But flake must preserve alpha channel in all operations.
Right now we'll loose alpha channel in this operation:
```nix
nix-repl> hex.darken (hex.alpha "#aaaaaa" 50) 50
"#1F1F1F"
```
There is obvious mistake in this example, so we can just say to user don't do things like this. But there can be a more difficult situation when user define color in one place in the label and modify it in another place so that's hard to follow up on this.
I think, that `{L, C, h}` also must has `a` field that means alpha channel. So `hexToOklch` must read alpha channel and `oklchToHex` write it to hex. All other operations must handle alpha channel too.
Some examples:
```nix
lighten = { L, C, h, a }@lch: percent:
    with lch; {
      L = lch.L + (percent / 100.0);
      inherit C h a; #alpha channel saved
    };

gradient = { L, C, h, a }@lch: { L, C, h, a }@another: steps:
    map (mod: {
      L = (1 - mod) * lch.L + mod * another.L;
      C = (1 - mod) * lch.C + mod * another.C;
      h = (1 - mod) * lch.h + mod * another.h;
      a = (1 - mod) * lch.a + mod * another.a; #alpha channel is gradiented too
    }) (modifiers steps);
```

Also I think that there must be not just `hex.alpha` function but `[hex.]adjustAlpha` (on percent) and `[hex.]setAlpha` functions. And exclusively for `hex` there must be `hex.stripAlpha` function that just uses `builtns.substring` function and cuts alpha channel information.

So what do you think about this? Isn't this overcomplicated?

nutsalhan87 commented:
Sorry, closed pr mistakenly

musjj commented:
Good point, I'll see what I can do here :thinking:

musjj commented:
Makes sense, I'll give it try!

musjj committed:
refactor conversion and alpha functions

musjj commented:
Ok, I added a few more functions for converting to/from sRGB and added the alpha functions you asked for. Does it look sound?

Btw, I just noticed that your `utils.round` is a round-half-down kind, is there a reason why you're not using a round-half-up?

nutsalhan87 commented:
Great work! You've also refactored other functions, thank you.
But I have small note. Putting converting to/from srgb logic into separate functions sounds convenient. But should they be in the main namespace? There no other functions which can handle `{ r, g, b }` set, only converting to hex or oklch. So maybe put them outside, in the `let .. in;` block or even in the `utils.nix`? In my opinion, they should be in the `utils.nix`. 
By the way, I want to refactor project structure because it's a mess right now. So you don't have to move to/from rgb functions outside the namespace -- I'll do this later. Of course, that's only if you agree that it's not necessary for them to be in the main namespace.
About round. 5.5 should round to 6. So, that is just my mistake. As I see, changing `(x - floored) > (ceiled - x)` to `>=` will fix this. If you want, you can change it yourself, or I will do this later.

musjj committed:
remove optional arguments for now

musjj commented:
Ok, makes sense, I'll give it a go.

Btw, would you mind exposing `utils.nix` as public API? There are some nice math functions in it that I think would be pretty useful for users.

musjj committed:
round half up instead of round half down

musjj commented:
Ok I've changed the rounding function. I've also made the `a` attribute as required now, because the optional binding didn't actually work unfortunately.

I've leave the `utils.nix` refactoring to you if that's okay. Some of the conversion functions depends on `clamp` which is not in `utils.nix` but in the top `let` block of `oklch.nix`, so I'm not completely sure how they should be re-organized.

nutsalhan87 commented:
I don't think that exposing `utils.nix` make sense because this flake is for color adjusting and it's strange that it also provides math functions. Maybe best way is to create own flake for math? But someone already created nix-math flake. By the way I don't remember why I don't use it: maybe I didn't know about it at the time of nix-colorizer creation or maybe it has some problems for me.

I think I'll put `clamp` into the `utils.nix`. It seems that function should be there, but I don't remember why it's not there.

I also want to make some changes in the main flake structure and want to discuss it. Now it looks like this:
```nix
{
  hexToOklch = ...;
  oklchToHex = ...;
  oklchsToHexes = ...;
  analogous = ...;
  complementary = ...;
  darken = ...;
  gradient = ...;
  lighten = ...;
  polygon = ...;
  shades = ...;
  splitComplementary = ...;
  tints = ...;
  tones = ...;
  hex = { ... };
}
``` 
But maybe it will be better to look like this:
```nix
{
  oklch = {
    to = {
      hex = ...;
      hexes = ...;
    };
    ... # color functions
  };
  hex = {
    to = {
      oklch = ...;
      oklchs = ...;
    };
    ... # color functions
  };
}
```

This structure looks more convenient. It's easier to add now color modes in the future. For example, to add srgb:
```nix
{
  oklch.to.srgb = ...;
  hex.to.srgb = ...;
  srgb = {
    to = {
      hex = ...;
      oklch = ...;
    };
    ... # color functions
  };
}
```

musjj commented:
Oooh yeah, I like that. It looks pretty ergonomic!

Btw, there's one more function I'm itching to add: `blend`. It's like `gradient`, but you specify a percentage between the two colors. It's useful when you just want to produce a blend between two colors, instead of a gradient array.

Do you think I could sneak it into this PR before your refactor work or should I do it in a separate PR?

nutsalhan87 commented:
Yes, you can push new function into this PR, why not

musjj committed:
add blend function for blending between two colors

musjj commented:
Ok, I added the function!
